src/metabase/pulse/render/body.clj

```diff
@@ -671,66 +634,100 @@
+        {:name          column-name
+         :cardName      card-name
          :type          card-type
          :data          selected-row-group
-         :yAxisPosition y-axis-pos}))))
+         :yAxisPosition y-axis-pos
+         :column        (second x-cols)
+         :breakoutValue group-key}))))
+
+(defn- render-multiple-lab-chart
+  "When multiple non-scalar cards are combined, render them as a line, area, or bar chart"
+  [render-type card dashcard {:keys [viz-settings]
+                              :as   data}]
+  (let [viz-settings      (merge viz-settings (:visualization_settings dashcard))
+        multi-res         (pu/execute-multi-card card dashcard)
+        ;; multi-res gets the other results from the set of multis.
+        ;; we shove cards and data here all together below for uniformity's sake
+        viz-settings      (set-default-stacked viz-settings card)
+        viz-settings-seqs (cons viz-settings (map (comp :visualization_settings :card) multi-res))
+        cards             (cons card (map :card multi-res))
+        multi-data        (cons data (map #(get-in % [:result :data]) multi-res))
+        axis-row-fns      (fn [card data]
+                            [(or (ui-logic/mult-x-axis-rowfn card data) #(vector (first %)))
+                             (or (ui-logic/mult-y-axis-rowfn card data) #(vector (second %)))])
+        rowfns            (mapv axis-row-fns cards multi-data)
+        row-seqs          (for [[row-seq [x-rowfn y-rowfn]] (map vector (map :rows multi-data) rowfns)]
+                            (map (comp flatten (juxt x-rowfn y-rowfn))
+                                 (common/row-preprocess x-rowfn y-rowfn row-seq)))
+        col-seqs          (map :cols multi-data)
+        first-rowfns      (first rowfns)
+        [[x-col] [y-col]] ((juxt (first first-rowfns) (second first-rowfns)) (first col-seqs))
+        labels            (x-and-y-axis-label-info x-col y-col viz-settings)
+        names             (map :name cards)
+        types             (replace {:scalar :bar} (map :display cards))
+        settings          (->ts-viz x-col y-col labels viz-settings)
+        y-axis-positions  (take (count names) (default-y-pos data axis-group-threshold))
+        series            (for [[enforced-type rows cols viz-settings [x-axis-rowfn y-axis-rowfn] y-axis-position card-name]
+                                (map vector types row-seqs col-seqs viz-settings-seqs rowfns y-axis-positions names)]
+                            (let [
+                                  x-cols      (x-axis-rowfn cols)
+                                  y-cols      (y-axis-rowfn cols)
+                                  x-rows      (map x-axis-rowfn rows)
+                                  y-rows      (map y-axis-rowfn rows)
+                                  joined-rows (map vector x-rows y-rows)]
+                              (map
+                               #(assoc % :yAxisPosition y-axis-position)
```

Comment:
This creates 2 helper fns to try keep `render-multiple-lab-chart` a bit more readable.

The way you have written it, there's actually a decent bit of pulling keys out  (eg. `(map :name cards)`), which, if you follow back to where the data comes from in the first place, you're using the same source. So, that was my thinking:

1. for each binding in the `for` (enforced-type rows cols viz-settings [x-axis-rowfn y-axis-rowfn] y-axis-position card-name), what are they coming from? Eg, `enforced-type` comes from the 'types' seq, which is just `(map :display cards)`, do this exercise for each of those bindings.
2. For every one of those seqs, find out where _they_ come from (that is, where does 'cards' come from?)
3. Do this upwards until you discover the common element(s). I originally tried to make the helper take `card` so I could map over cards, but some of what was happening in the `for` actually needed `data`, which comes from the results
4. Figure out how to get the sequence of this 'topmost data', and build a helper that you can map over THAT instead of mapping over a bunch of seqs.

Reply:
ALSO, please see if the tests actually pass with my suggested changes above. One thing I wasn't sure of before the end of the day was if `joined-rows` (inside my `card-result->series` fn) is in the correct shape.

Reply:
Thanks for the thorough explanation! This really helps me understand the reason behind the refactor.

I think I could adjust `joined-rows` a bit, if I'm not mistaken this is almost correct.

I also agree that working with each individual `card` and `data` is a ton easier than trying to work on lots of seqs/lists